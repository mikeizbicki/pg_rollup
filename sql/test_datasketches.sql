create or replace language plpython3u;
create extension if not exists hll;
create extension if not exists pg_rollup;
create extension if not exists datasketches;

CREATE OR REPLACE FUNCTION kll_float_sketch_union(a kll_float_sketch, b kll_float_sketch) RETURNS kll_float_sketch AS $$
    select kll_float_sketch_merge(sketch) from (select a as sketch union all select b) t;
$$ LANGUAGE 'sql' IMMUTABLE PARALLEL SAFE;

create table dstest (
    id serial primary key,
    a int
);

insert into dstest (a) (select * from generate_series(0,10000));

select create_rollup(
    'dstest',
    'dstest_rollup1',
    rollups => $$
        kll_float_sketch(a)
    $$
);

select create_rollup(
    'dstest',
    'dstest_rollup2',
    rollups => $$
        theta_sketch(a)
    $$
);

-- FIXME:
-- kll_float_sketch is non-deterministic due to an internal random number generation;
-- this causes the groundtruth and the rollup table to diverge slightly,
-- and so we can't test for equality

--select assert_rollup('dstest_rollup1');
select assert_rollup('dstest_rollup2');

insert into dstest (a) (select * from generate_series(0,10000));

--select assert_rollup('dstest_rollup1');
select assert_rollup('dstest_rollup2');

