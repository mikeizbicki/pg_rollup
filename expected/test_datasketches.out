/*
create or replace language plpython3u;
create extension if not exists hll;
create extension if not exists pg_rollup;
create extension if not exists datasketches;

CREATE OR REPLACE FUNCTION kll_float_sketch_union(a kll_float_sketch, b kll_float_sketch) RETURNS kll_float_sketch AS $$
    select kll_float_sketch_merge(sketch) from (select a as sketch union all select b) t;
$$ LANGUAGE 'sql' IMMUTABLE PARALLEL SAFE;

CREATE OR REPLACE FUNCTION frequent_strings_sketch_union(a frequent_strings_sketch, b frequent_strings_sketch) RETURNS frequent_strings_sketch AS $$
    select frequent_strings_sketch_merge(9,sketch) from (select a as sketch union all select b) t;
$$ LANGUAGE 'sql' IMMUTABLE PARALLEL SAFE;

create table dstest (
    id serial primary key,
    a int
);

insert into dstest (a) (select * from generate_series(0,10000));

select create_rollup(
    'dstest',
    'dstest_rollup1',
    rollups => $$
        kll_float_sketch(a)
    $$
);

select create_rollup(
    'dstest',
    'dstest_rollup2',
    rollups => $$
        frequent_strings_sketch(text(a))
    $$
);

-- FIXME:
-- kll_float_sketch is non-deterministic due to an internal random number generation;
-- this causes the groundtruth and the rollup table to diverge slightly,
-- and so we can't test for equality

--select assert_rollup('dstest_rollup1');
select assert_rollup('dstest_rollup2');

insert into dstest (a) (select * from generate_series(0,10000));

--select assert_rollup('dstest_rollup1');
select assert_rollup('dstest_rollup2');
*/
